### PR TITLE
code-generator: Go native codegen for gen-helpers

### LINKED ISSUE #163: Allow to use the code-gen in vendorless projects
## Background

At Knative project, we would like to get rid of the `vendor` directory (see knative/infra#134 for details). One of the things that block that is the extensive use of the Kubernetes code-generator, to be exact, the [generate-groups.sh](https://github.com/kubernetes/code-generator/blob/3ba1295f0ae75f07740f3b2d33c8af163881cbf3/generate-groups.sh) shell script.

Recently, the shell code in the Kubernetes has been rewritten a bit in the right direction (https://github.com/kubernetes/kubernetes/pull/117262). 

However, it still depends on the non-go shell scripts as the entry points, thus still requiring the use of `vendor` in the code. The usage of shell scripts isn't cross-platform.

## Expected behavior

The Kubernetes code-gen should allow projects to leverage the `go run ...` syntax. With it, projects could use the code-gen even without any connection to Kuberentes code:

```
$ go run k8s.io/code-generator@v0.28.1 deepcopy \
  knative.dev/pkg/client knative.dev/pkg/apis \
  'duck:v1alpha1,v1beta1,v1' \
  --go-header-file "${REPO_ROOT_DIR}/hack/boilerplate/boilerplate.go.txt"
```

Or by managing the code-gen version in the project's `go.mod` file:

```go
module knative.dev/serving

go 1.20

require (
	k8s.io/code-generator v0.28.1
)
```

and then running withing that project (notice the lack of the package version):

```
$ go run k8s.io/code-generator deepcopy \
  knative.dev/pkg/client knative.dev/pkg/apis \
  'duck:v1alpha1,v1beta1,v1' \
  --go-header-file "${REPO_ROOT_DIR}/hack/boilerplate/boilerplate.go.txt"
```

### Refactor into an entry point binary.

The K8s code-gen already has multiple binaries in the [`cmd/` directory](https://github.com/kubernetes/code-generator/tree/3ba1295f0ae75f07740f3b2d33c8af163881cbf3/cmd), but we could add the entry point `code-gen` binary there, that will directly support the same arguments as the current [kube_codegen.sh](https://github.com/kubernetes/code-generator/blob/3ba1295f0ae75f07740f3b2d33c8af163881cbf3/kube_codegen.sh) script. This approach is cross-platform.

### Embed the shell script in Go

An easier solution would be to use an approach similar to knative/hack#222, so embedding the shell scripts into Go code (with `//go:embed` feature), and extracting it into some `tmp/` directory.

### PR DESCRIPTION
**Pull request by cardil**
#### What type of PR is this?

/kind feature
/kind cleanup

#### What this PR does / why we need it:

This PR rewrites the shell code from `kube_codegen.sh` into a Go code with both executable, and packages which may be used downstream. I purposely limited the scope for now to just `kube::codegen::gen_helpers`. The follow-up PRs are, of course, expected.

The reasons for this are explained in the connected issue: https://github.com/kubernetes/code-generator/issues/163.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/code-generator/issues/163 (partially)

#### Special notes for your reviewer:

This is the first PR for this issue. Expect PRs for other shell functions: `kube::codegen::gen_openapi` and `kube::codegen::gen_client`.

The idea is not to change the logic of the generator, or the flags, and to keep 100% compatibility.

#### Does this PR introduce a user-facing change?

```release-note
The Kubernetes code-generator has now the new Go-native entrypoint wrapper. This allows for easier management in downstream projects, and usage with Go' `go run k8s.io/code-generator` command, even without the Kubernetes sources.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by cardil** [comment]
Need to fix the tests. There are failing on CI, and passing locally... Setting `[WIP]`

File: staging/src/k8s.io/code-generator/cmd/conversion-gen/args/args.go [hunk 1]
```diff
@@ -68,13 +68,13 @@ func NewDefaults() (*args.GeneratorArgs, *CustomArgs) {
 
 // AddFlags add the generator flags to the flag set.
 func (ca *CustomArgs) AddFlags(fs *pflag.FlagSet) {
-	pflag.CommandLine.StringSliceVar(&ca.BasePeerDirs, "base-peer-dirs", ca.BasePeerDirs,
+	fs.StringSliceVar(&ca.BasePeerDirs, "base-peer-dirs", ca.BasePeerDirs,
```
Comment (thockin) [inline staging/src/k8s.io/code-generator/cmd/conversion-gen/args/args.go:71]: You could send fixes like this as a PR of its own

**Comment by thockin** [comment]
A few thoughts.

1) I am deep in the midst of retooling ALL of these with what is essentially a different form of gengo.  As such, our changes collide in a BIG way.  It would help me to not lose my mind if we didnt try to do both at once.

2) My changes should also solve the BIZARRE way these tools engage with vendor/ and GOPATH.  Can we maybe work together to see if that solves it all at once?

File: staging/src/k8s.io/code-generator/cmd/conversion-gen/args/args.go (hunk 1 above)
Comment (cardil) [reply staging/src/k8s.io/code-generator/cmd/conversion-gen/args/args.go:71]: This is, in fact, crucial to the implementation. Well, the tests of the implementation. 

I'm calling the specific generator function from the main binary. Without this, the tests fail, as I'm unable to pass proper flagset here.

What I can do, is to send this fix earlier, to be merged before this refactor, in a separate PR.

**Comment by cardil** [comment]
> A few thoughts.

Thanks for looking into this. I appreciate that!

> 1. I am deep in the midst of retooling ALL of these with what is essentially a different form of gengo.  As such, our changes collide in a BIG way.  It would help me to not lose my mind if we didnt try to do both at once.

Would you mind sending me your changes to compare? So, I would know what we are talking about? I presume the most colliding thing is the changes I made to the `cmd/` main files, by extracting reusable, embeddable functions out of them. But, this is just a guess. I would love to see the exact changes, or at least your idea of the retooling them in depth. 

> 2. My changes should also solve the BIZARRE way these tools engage with vendor/ and GOPATH.  Can we maybe work together to see if that solves it all at once?

I think, I'm hitting this `vendor/` and `GOPATH` issue, myself as well. I wrote [a kinda integration test `generator_test.go`](https://github.com/cardil/kubernetes/blob/7244b96325f91f8fe99761a256cbb3a3b9a10946/staging/src/k8s.io/code-generator/pkg/codegen/helpers/generator_test.go), to assert that the new wrapper binary, doesn't change a bit in a generated code (in `examples` folder of code-generator). This test pass nicely when executed normally:

```sh
$ go test -count=1 k8s.io/code-generator/pkg/codegen/helpers
ok  	k8s.io/code-generator/pkg/codegen/helpers	5.544s
```

But it fails when called via Makefile:

```sh
$ make test WHAT=./vendor/k8s.io/code-generator/pkg/codegen/helpers
go version go1.21.5 linux/amd64
+++ [1221 14:33:25] Set GOMAXPROCS automatically to 16
+++ [1221 14:33:25] Running tests without code coverage and with -race
--- FAIL: TestGenerate (0.57s)
[... cut for brevity ...]
    klog.go:897: I1221 14:33:26.734056] conversion-gen completed successfully
    klog.go:897: I1221 14:33:26.734090] Successfully generated helpers for k8s.io/code-generator/examples
    generator_test.go:76: got "ff4ebd9370efdcc238472782a872d03b4dca8cb40a82ba30c6929db92459f757", want "8aeaab5a8531a157cb81a6122523ee0939539a8839d21536ff7a11492a4e980a"
FAIL
FAIL	k8s.io/kubernetes/vendor/k8s.io/code-generator/pkg/codegen/helpers	0.583s
FAIL
make: *** [Makefile:191: test] Error 1
```

The fail is there because, for some unknown to me reason, the gengo doesn't generate code from that Makefile setup.

I would love to work together. Maybe we could set up a call after the holidays to disqus the approach. I also wouldn't like to work on the same exact competing changes. I'm looking forward to that, I think we could help each other a lot!

**Comment by cardil** [comment]
> A few thoughts.

Thanks for looking into this. I appreciate that!

> 1. I am deep in the midst of retooling ALL of these with what is essentially a different form of gengo.  As such, our changes collide in a BIG way.  It would help me to not lose my mind if we didnt try to do both at once.

Would you mind sending me your changes to compare? So, I would know what we are talking about? I presume the most colliding thing is the changes I made to the `cmd/` main files, by extracting reusable, embeddable functions out of them. But, this is just a guess. I would love to see the exact changes, or at least your idea of the retooling them in depth. 

> 2. My changes should also solve the BIZARRE way these tools engage with vendor/ and GOPATH.  Can we maybe work together to see if that solves it all at once?

I think, I'm hitting this `vendor/` and `GOPATH` issue, myself as well. I wrote [a kinda integration test `generator_test.go`](https://github.com/cardil/kubernetes/blob/7244b96325f91f8fe99761a256cbb3a3b9a10946/staging/src/k8s.io/code-generator/pkg/codegen/helpers/generator_test.go), to assert that the new wrapper binary, doesn't change a bit in a generated code (in `examples` folder of code-generator). This test pass nicely when executed normally:

```sh
$ go test -count=1 k8s.io/code-generator/pkg/codegen/helpers
ok  	k8s.io/code-generator/pkg/codegen/helpers	5.544s
```

But it fails when called via Makefile:

```sh
$ make test WHAT=./vendor/k8s.io/code-generator/pkg/codegen/helpers
go version go1.21.5 linux/amd64
+++ [1221 14:33:25] Set GOMAXPROCS automatically to 16
+++ [1221 14:33:25] Running tests without code coverage and with -race
--- FAIL: TestGenerate (0.57s)
[... cut for brevity ...]
    klog.go:897: I1221 14:33:26.734056] conversion-gen completed successfully
    klog.go:897: I1221 14:33:26.734090] Successfully generated helpers for k8s.io/code-generator/examples
    generator_test.go:76: got "ff4ebd9370efdcc238472782a872d03b4dca8cb40a82ba30c6929db92459f757", want "8aeaab5a8531a157cb81a6122523ee0939539a8839d21536ff7a11492a4e980a"
FAIL
FAIL	k8s.io/kubernetes/vendor/k8s.io/code-generator/pkg/codegen/helpers	0.583s
FAIL
make: *** [Makefile:191: test] Error 1
```

The fail is there because, for some unknown to me reason, the gengo doesn't generate code from that Makefile setup.

I would love to work together. Maybe we could set up a call after the holidays to disqus the approach. I also wouldn't like to work on the same exact competing changes. I'm looking forward to that, I think we could help each other a lot!

**Comment by thockin** [comment]
I tried it one way (https://github.com/thockin/kubernetes/tree/go-workspaces-plus-vendor) but found I made a mess and it became hard to review.

So I started again (https://github.com/thockin/kubernetes/tree/go-workspaces-except-gengo).

The core problem is that gengo assumes (all over the place, in subtle ways) that directories and import-paths are interchangeable and that you can string-cat and trim-prefix your way between them - you cannot.  We set up staging in such a way that it KIND OF works but you have to invoke all of the tools in exactly the right way or they fails in a hundred ways with vague and unhelpful errors.  Or worse, they don't fail but writes files wherever they like.

My branch basically forks gengo-v2, replaces all the manual load-and-parse logic with Go's own `packages.Load` and I am in the midst of retooling each gengo tool to distinguish dirtectory paths from import paths.  The net result is that the tools will be much more flexible, and will take any Go "pattern" (k8s.io/repo/dir or ./staging/src/repo/dir or even ./vendor/site/repo/dir, as well as the respective /... forms).  All of the GOPATH and vendor special-casing will go away, and hopefully gengo/v2 get thinner and thinner, until it barely exists.

**Comment by cardil** [comment]
Your description makes 100% sense. I'll dig into details as soon as I get some time.

Thanks very much for the links! :heart:

**Comment by thockin** [comment]
At the end of this, we may still want to replace or wrap the N tools with one tool (as kube-codegen.sh does), but that's a tomorrow-problem :)

**Comment by cardil** [comment]
Your description makes 100% sense. Furthermore, it explains the failures I see (from the mentioned above test). It's clearly visible that the problems are the package imports that couldn't be found in that phony GOPATH setup (still got a hard time understanding how exactly it is configured):

```
    klog.go:897: I1221 14:33:26.732939] importPackage k8s.io/code-generator/examples/apiserver/apis/example/v1
    klog.go:897: I1221 14:33:26.732967] addDir k8s.io/code-generator/examples/apiserver/apis/example/v1
    klog.go:897: I1221 14:33:26.733226] importPackage k8s.io/code-generator/examples/apiserver/apis/example2/v1
    klog.go:897: I1221 14:33:26.733251] addDir k8s.io/code-generator/examples/apiserver/apis/example2/v1
    klog.go:897: I1221 14:33:26.733506] importPackage k8s.io/code-generator/examples/apiserver/apis/example3.io/v1
    klog.go:897: I1221 14:33:26.733530] addDir k8s.io/code-generator/examples/apiserver/apis/example3.io/v1
    klog.go:897: I1221 14:33:26.733756] importPackage k8s.io/apimachinery/pkg/runtime
    klog.go:897: I1221 14:33:26.733779] addDir k8s.io/apimachinery/pkg/runtime
```


I'll dig into details as soon as I get some time. But, my current understanding is that first the gengo/v2 should land and then something like this PR could be merged. I'm fine with that, and I could help with it.

Thanks very much for the links! :heart:

**Comment by cardil** [comment]
I've been looking a bit your changes at https://github.com/kubernetes/kubernetes/compare/master...thockin:kubernetes:go-workspaces-except-gengo, but it's not obvious to me how I could help you with that effort. But, such rewrite should definitely go in before change like this PR.

Maybe you could think of a way to divide your refactor into smaller chunks, so we can contribute better.

**Comment by cardil** [comment]
@thockin I've been looking a bit your changes at https://github.com/kubernetes/kubernetes/compare/master...thockin:kubernetes:go-workspaces-except-gengo, but it's not obvious to me how I could help you with that effort. But, such rewrite should definitely go in before change like this PR.

Maybe you could think of a way to divide your refactor into smaller chunks, so we can contribute better.

**Comment by thockin** [comment]
In theory "help" could mean any of:
- tackle items from the TODO list (which I have on my local machine, but am happy to share)
- review
- help make it more reviewable
- add tests (gengo has none)
- test manually and find things I missed

**Comment by cardil** [comment]
How about we start by creating an epic in https://github.com/kubernetes/code-generator/issues to track the whole effort. That would be a good place to share the TODO you have, discuss things, and plan work to be more reviewable.

**Comment by cardil** [comment]
Hey @thockin! Today, I figured out that my changes here, and yours from https://github.com/kubernetes/kubernetes/compare/master...thockin:kubernetes:go-workspaces-except-gengo are mostly orthogonal, and shouldn't block each other.

I'm focusing here, only, on the refactoring of the `gen_*` shell functions from the https://github.com/kubernetes/kubernetes/blob/9bd0ef5f173de3cc2d1d629a4aee499d53690aee/staging/src/k8s.io/code-generator/kube_codegen.sh file. And, you are changing the whole setup of building and testing the whole project. That's great, and as I said, I can help you with that, if you can divide the work into maintainable chunks, and act as mini-project leader.

However, I think your work shouldn't block my changes.

Today, I've checked, how much effort would it bring, for you, the merging of this PR. See the conflicts here: https://github.com/thockin/kubernetes/pull/14/conflicts. As far as I can see, the conflicts aren't very problematic, and resolving them should be straightforward. Even more, I can file a second, smaller PR, as already discussed, that will refactor just the `cmd`'s `main()` function split into `embed` subpackages.

WDYT?

**Comment by cardil** [comment]
> Hey @thockin! Today, I figured out that my changes here, and yours from [master...thockin:kubernetes:go-workspaces-except-gengo](https://github.com/kubernetes/kubernetes/compare/master...thockin:kubernetes:go-workspaces-except-gengo) are mostly orthogonal, and shouldn't block each other.
> 
> I'm focusing here, only, on the refactoring of the `gen_*` shell functions from the https://github.com/kubernetes/kubernetes/blob/9bd0ef5f173de3cc2d1d629a4aee499d53690aee/staging/src/k8s.io/code-generator/kube_codegen.sh file. And, you are changing the whole setup of building and testing the whole project. That's great, and as I said, I can help you with that, if you can divide the work into maintainable chunks, and act as mini-project leader.
> 
> However, I think your work shouldn't block my changes.
> 
> Today, I've checked, how much effort would it bring, for you, the merging of this PR. See the conflicts here: https://github.com/thockin/kubernetes/pull/14/conflicts. As far as I can see, the conflicts aren't very problematic, and resolving them should be straightforward. Even more, I can file a second, smaller PR, as already discussed, that will refactor just the `cmd`'s `main()` function split into `embed` subpackages.
> 
> WDYT?

Gentle ping

**Comment by cardil** [comment]
/test pull-kubernetes-verify

Looks like a flaky test?

**Comment by cardil** [comment]
@upodroid @thockin The PR is ready for reviewing.

**Comment by thockin** [comment]
Given https://kubernetes.slack.com/archives/C0EG7JC6T/p1705698354201949 discussion, this is on hold, right?

**Comment by cardil** [comment]
> Given https://kubernetes.slack.com/archives/C0EG7JC6T/p1705698354201949 discussion, this is on hold, right?

Yes. On hold. Likely to be closed, soon.

**Pull request closed without merging**